src/commands.js

```diff
@@ -0,0 +1,183 @@
+var Q = require("q");
+var _ = require("lodash");
+
+const VARIABLE = /^\{(int)(text)|:[a-zA-Z]+\}$/;
+const CHOICE_VARIABLE = /^\{([a-zA-Z]+\|?)+\}$/;
+const CONSTANT_VALUE = /^[a-zA-Z]+$/;
+
+const VARIABLE_TYPES = {
+    "TEXT": {
+        match_expression: /text/,
+        parse: function(token){
+            return token; //all tokens are valid
+        }
+    },
+    "INT": { 
+        match_expression: /int/, 
+        parse: function(token){
+            var value = parseInt(token, 10);
+            if(isNaN(value)) throw new InvalidTypeValueError(token, "int");
+            return value;
+        }
+    }
+    /*...*/
+};
+
+function InvalidTypeValueError(token, type){
+    this.token = token;
+    this.type = type;
+    this.name = 'InvalidTypeValueError';
+    this.stack = (new Error()).stack;
+}
+
+/* 
+ * This command tokenizer takes the following token descriptions
+ * constant: constantValue - constant values will be varified to be the exact 
+ *           value specified. If it does not match, this function will throw 
+ *           InvalidConstantError
+ * variable: {type: variableName} - variable values will be varified to be the 
+ *           type specified. If the type does not match, this function will throw
+ *           InvalidTypeValueError
+ * choice:   {option1|option2|option3} - choices must be one of the prespecified  
+ *           values. The option will be stored in the key and value of the 
+ *           returned token object. If it does not match one of the specified 
+ *           options, this function will throw InvalidChoiceError
+ * 
+ * types:    int, text - type must match one of these. If it does not match, thi
+ *           this function will throw InvalidTypeError
+ * Commands that are too short will throw TooFewTokensError
+ */
+function tokenize(command, descriptions){
+    return Q.fcall(function(){
+        var tokens = getTokens(_.toLower(command), descriptions.length);    
+
+        var parameters = {};
+        _.zipWith(tokens, descriptions, function(token, description){
+            if(_.isNil(token)){
+                throw new TooFewTokensError(tokens, descriptions);
+            }else if(description.match(VARIABLE)){
+                parameterizeVariableToken(token, description, parameters);
+            }else if(description.match(CHOICE_VARIABLE)){
+                parameterizeChoiceToken(token, description, parameters);
+            }else if(description.match(CONSTANT_VALUE)){
+                parameterizeConstantToken(token, description, parameters); 
+            }else{
+                throw new InvalidTokenDescriptionError(description);
+            }
+        });
+        return parameters;
+    });
+}
+
+function TooFewTokensError(tokens, descriptions){
```

Comment:
Yeah - I did that for convenience at first but should have relocated them after. I will move them.